xdsl/parser/attribute_parser.py

```diff
@@ -504,21 +504,48 @@ def parse_shape_delimiter(self) -> None:
         # Move the lexer to the position after 'x'.
         self._resume_from(self._current_token.span.start + 1)
 
+    def _parse_empty_dimension_list(self) -> bool:
+        """
+        Parses an empty dimension list, which has it's own syntax:
+          empty-dimension-list ::= `[` `]`
+        Note that a regular dimension list has no square brackets.
+        """
+        if self.parse_optional_characters("["):
+            self.parse_characters("]")
+            return True
+
+        return False
+
+    def parse_dimension_list(self) -> list[int]:
+        """
+        Parse a dimension list with the following format:
+          dimension-list ::= (dimension `x`)* dimension `x`?
+        each dimension is also required to be non-negative.
+        """
+        dims: list[int] = []
+        accepted_token_kinds = (MLIRTokenKind.INTEGER_LIT, MLIRTokenKind.QUESTION)
+
+        if self._parse_empty_dimension_list():
+            return []
+
+        while self._current_token.kind in accepted_token_kinds:
+            dim = self.parse_shape_dimension()
+            dims.append(dim)
+
+            # on last iteration, 'x' may not exist, e.g.: 5x3x2
+            if len(self._current_token.text) > 0 and self._current_token.text[0] == "x":
+                self.parse_shape_delimiter()
+
+        return dims
```

Comment:
Would something like this work?

```
dims = []

while token is int or ?:
    dims.append(parse_dim())
    if not parse_optional_shape_delimiter():
        break

return dims
```

This way we consume the trailing x if there is one, and limit the shape delimiter logic to one function.